### PR TITLE
added evalQueue.await to the iodide api

### PR DESCRIPTION
**Pull request by hamilton**
This PR adds another feature to the iodide api - `iodide.evalQueue.await`. Motivated by the discussion in #597, we want an easy way to prevent further cell evaluation until certain Promises are resolved, specifically any using the `fetch` api. So `iodide.evalQueue.await` works exactly as `Promise.all` - you pass in an array of promises (like `fetch` calls) and it waits for all of the Promises to resolve / reject before moving on.

I would imagine that the majority of data scientists / our users would prefer to just use this function instead of `iodide.evalQueue.requireExplicitContinuation` / `iodide.evalQueue.continue`, but those pieces are functionality will be vital in non-Promise situations, and at any rate I use both of them within `iodide.evalQueue.await`.

Use `?gist=hamilton/50e813eb8169f0486b48d55d5d206df8` for an example.  Below is a code block that demonstrates the functionality:

```javascript
let url1 = 'https://normandy.cdn.mozilla.net/api/v2/recipe/?format=json&ordering=-last_updated'
let url2 = 'https://normandy.cdn.mozilla.net/api/v2/recipe/?format=json&ordering=-last_updated&page=2'
var data1 =[]
var data2 = []

let timer = new Promise((go)=>{
  setTimeout(()=>{
    go('setTimeout Resolved.')
  }, 4000)
})

iodide.evalQueue.await([
  timer,
  fetch(url1).then(d=>d.json()).then(d=>{data1 = d; return d}),
  fetch(url2).then(d=>d.json()).then(d=>{data2 = d; return d})
])

```

### CODE REVIEW

**Comment by mdboom** [comment]
Cool.  This looks like it will help the existing `pyodide.loadPackage` functionality work with "run all".  

What happens if this is called multiple times in the same cell?  It might be called from within a called function, for example, and not actually be present in the cell content, therefore it might be easy for the user to accidentally call it twice.

**Comment by bcolloran** [comment]
looks very nice. ping when you're ready for review

**Comment by hamilton** [comment]
@mdboom + @bcolloran I've been playing with some ideas around how we want to handle this. I've migrated over to a version in which the api calls increment + decrement a new cell field `asyncProcessCount`. The polling then watches that counter, and once it hits zero, resolves. This allows multiple / nested calls for `iodide.evalQueue.await`, `iodide.evalQueue.requireExplicitContinuation` (which increments) and `iodide.evalQueue.continue` (which decrements).

This feels like the right way to move forward with this, api-wise. Let me know what y'all think.

**Comment by bcolloran** [comment]
needs some error handling. for example, this prevents execution of further js cells:

```
let url1 = 'https://normandy.cdn.mozilla.net/api/badrequest'
let url2 = 'https://normandy.cdn.mozilla.net/api/v2/recipe/?format=json&ordering=-last_updated&page=2'
var data1 =[]
var data2 = []

let timer = new Promise((go)=>{
  setTimeout(()=>{
    go('setTimeout Resolved.')
  }, 4000)
})

iodide.evalQueue.await([
  timer,
  fetch(url1).then(d=>d.json()).then(d=>{data1 = d; return d}),
  fetch(url2).then(d=>d.json()).then(d=>{data2 = d; return d})
])
```

api looks great.

code looks ok to the extent i can follow it, but i don't know promise stuff that well, so requesting code review from @wlach

**Comment by hamilton** [comment]
@bcolloran thanks for that catch. I've added more error handling around `evalQueue.await`. Any rejection or error will now be caught in a `.catch` in the api, so we can surface the error message and free up further cell evaluation (we can revisit THAT part down the line)

File: src/iodide-api/evalQueue.js [hunk 1]
```diff
@@ -34,7 +59,29 @@ export const waitForExplicitContinuationStatusResolution = () => new Promise((re
   }
 })
 
+const awaitPromises = promises =>
+  // is functionally identical to Promise.all.
+  Promise.resolve()
```
Comment (wlach) [inline src/iodide-api/evalQueue.js]: Is there a reason why you're creating a problem that is immediately resolved here? Could we not just key this off of Promises.all() and do whatever state setting we want to do immediately?

Comment (hamilton) [reply src/iodide-api/evalQueue.js]: yep this is madness, fixing

**Pull request merged**